controller/logs.go

```diff
@@ -76,9 +100,64 @@ func (c *Controller) LogsForDeployment(ctx context.Context, req *entity.Deployme
 				continue
 			}
 			// Output logs
-			fmt.Print(strings.Join(logDiff, "\n"))
+			fmt.Println(strings.Join(logDiff, "\n"))
+			// Set out walk pointer forward using the newest logs
+			prevLogs = currLogs
+		}
+	}
+	return nil
+}
+
+func (c *Controller) LogsForBuild(ctx context.Context, req *entity.DeploymentLogsRequest) error {
+	// Fetch Initial Deployment Logs
+	query := entity.DeploymentGQL{
+		BuildLogs: true,
+	}
+	deploy, err := c.gtwy.GetDeploymentByID(ctx, &entity.DeploymentByIDRequest{
+		DeploymentID: req.DeploymentID,
+		ProjectID:    req.ProjectID,
+		GQL:          query,
+	})
+	if err != nil {
+		return err
+	}
+	// Break them down line by line
+	logLines := strings.Split(deploy.BuildLogs, "\n")
+	offset := 0.0
+	if req.NumLines != 0 {
+		// If a limit is set, walk it back n steps (with a min of zero so no panics)
+		offset = math.Max(float64(len(logLines))-float64(req.NumLines)-1, 0.0)
+	}
+	// Output Initial Logs
+	fmt.Println(strings.Join(logLines[int(offset):], "\n"))
+	if req.NumLines == 0 {
+		// If no log limit is set, we stream logs
+		prevLogs := strings.Split(deploy.BuildLogs, "\n")
+		for {
+			time.Sleep(time.Second * 2)
+			deploy, err := c.gtwy.GetDeploymentByID(ctx, &entity.DeploymentByIDRequest{
+				DeploymentID: req.DeploymentID,
+				ProjectID:    req.ProjectID,
+				GQL:          query,
+			})
+			if err != nil {
+				return err
+			}
+			// Current Logs fetched from server
+			currLogs := strings.Split(deploy.BuildLogs, "\n")
+			// Diff logs using the line numbers as references
+			logDiff := currLogs[len(prevLogs)-1 : len(currLogs)-1]
+			// If no changes we continue
+			if len(logDiff) == 0 {
+				continue
+			}
+			// Output logs
+			fmt.Println(strings.Join(logDiff, "\n"))
 			// Set out walk pointer forward using the newest logs
 			prevLogs = currLogs
+			if deploy.Status != entity.STATUS_BUILDING {
+				break
+			}
```

Comment:
These two functions are SO similar and yet I can't seem to figure out a way to abstract the commonalities into something pretty :(

Reply:
Are the only differences `query`, `deploy.BuildLogs` and `STATUS_BUILDING` references? If so, I'd make a new fn with those as args, then have `LogsForBuild` and `LogsForDeploy` call into it.

Reply:
Mostly, but they also rely on the closure of having prevlogs memory pointer, as well as this for loop which is broken for build logs but not for deployment logs

Will continue a bit but just wanted to mention/get thoughts